### PR TITLE
Fix MemoryDebugView and override Memory.ToString similar to Span

### PR DESCRIPTION
**Pull request by ahsonkhan**
**TODO:** Mirror into corefx ~and add back ToString() to the ref + add tests.~

Related PR in corefx: https://github.com/dotnet/corefx/pull/27694

cc @jkotas, @KrzysztofCwalina, @stephentoub, @pakrym, @khellang

### CODE REVIEW

File: src/mscorlib/shared/System/Memory.cs [hunk 1]
```diff
@@ -163,6 +160,30 @@ private Memory(object obj, int index, int length)
...
+                {
+                    return str.Substring(_index, _length);
+                }
+                else if (_index < 0)
+                {
+                    return Span.ToString();
+                }
+                else if (_object != null)
+                {
+                    return new string((char[])_object);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/Memory.cs]: Is this correct?  Don't you also need to factor in _index and _length?

File: src/mscorlib/shared/System/ReadOnlyMemory.cs [hunk 2]
```diff
@@ -125,6 +122,30 @@ internal ReadOnlyMemory(object obj, int start, int length)
...
+                {
+                    return str.Substring(_index, _length);
+                }
+                else if (_index < 0)
+                {
+                    return Span.ToString();
+                }
+                else if (_object != null)
+                {
+                    return new string((char[])_object);
```
Comment (stephentoub) [inline src/mscorlib/shared/System/ReadOnlyMemory.cs]: Same question.

File: src/mscorlib/shared/System/Memory.cs [hunk 3]
```diff
@@ -163,6 +160,30 @@ private Memory(object obj, int index, int length)
...
 
+        /// <summary>
+        /// For <see cref="Memory{Char}"/>, returns a new instance of string that represents the characters pointed to by the memory.
+        /// Otherwise, returns a <see cref="string"/> with the name of the type and the number of elements.
+        /// </summary>
+        public override string ToString()
+        {
+            if (typeof(T) == typeof(char))
+            {
+                if (_object is string str)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Memory.cs]: Why can't this be just:

```
if (typeof(T) == typeof(char))
{
    return Span.ToString();
}
```

?

File: src/mscorlib/shared/System/Memory.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: Yep, overlooked that case. We need to support sliced memory.

File: src/mscorlib/shared/System/Memory.cs (hunk 3 above)
Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: I thought of doing that (I do that for the OwnedMemory case anyway), but in the case of string, we would create a span from the string, slice it, and and then grab the string again. Seemed unnecessary.

Similarly, with array, we would construct the span around the array, pin it, and create the new string. It was simpler to just create the substring directly from the array.

Span.ToString() would be a clean implementation though, if you think that is the way to go.

Comment (jkotas) [reply src/mscorlib/shared/System/Memory.cs]: > in the case of string, we would create a span from the string, slice it, and and then grab the string again.

Ok, this make sense as an optimization.

> t was simpler to just create the substring directly from the array.

This is not saving anything meaningful. Constructing Span from array is cheap. So what about something like:

```
if (typeof(T) == typeof(char))
{
     return (_object is string str) ? str.Substring(_index, _length)  : Span.ToString();
}
```

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: > This is not saving anything meaningful. Constructing Span from array is cheap.

Yes, that works. I will add and run tests on the corefx side to ensure correctness.

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: What about this extra check that we end up doing when _object is a char[] if we call Span.ToString() instead?
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Memory.cs#L215

Comment (jkotas) [reply src/mscorlib/shared/System/Memory.cs]: It costs nothing. You can always measure it if you are not sure.

Comment (ahsonkhan) [reply src/mscorlib/shared/System/Memory.cs]: Yep, just measured. It has no impact.

**Pull request merged**